src/parameterFiles.ts

```diff
@@ -0,0 +1,507 @@
+// ----------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ----------------------------------------------------------------------------
+
+import * as assert from 'assert';
+import * as fse from 'fs-extra';
+import * as path from 'path';
+import { commands, ConfigurationTarget, MessageItem, TextDocument, Uri, ViewColumn, window, workspace } from 'vscode';
+import { callWithTelemetryAndErrorHandling, DialogResponses, IActionContext, IAzureQuickPickItem, UserCancelledError } from 'vscode-azureextensionui';
+import { configKeys, configPrefix, globalStateKeys, isWin32 } from './constants';
+import { DeploymentTemplate } from './DeploymentTemplate';
+import { ext } from './extensionVariables';
+import { containsParamsSchema } from './schemas';
+
+const readAtMostBytesToFindParamsSchema = 4 * 1024;
+const currentMessage = "Current";
+const similarFilenameMessage = "Similar filename";
+const howToMessage = `You can manually associate a parameter file with this template at any time by clicking "Select Parameter File..." in the status bar or the editor context menu.`;
+
+const _filesToIgnoreThisSession: Set<string> = new Set<string>();
+
+interface IPossibleParamFile {
+  uri: Uri;
+  friendlyPath: string;
+  isCloseNameMatch: boolean;
+}
+
+// tslint:disable-next-line: max-func-body-length
+export async function selectParameterFile(actionContext: IActionContext, sourceUri: Uri | undefined): Promise<void> {
+  if (!sourceUri) {
+    sourceUri = window.activeTextEditor?.document.uri;
+  }
+  if (!sourceUri) {
+    await ext.ui.showWarningMessage(`No Azure Resource Manager template file is being edited.`);
+    return;
+  }
+
+  let templateUri: Uri = sourceUri;
+
+  // Verify it's a template file
+  const contents = (await fse.readFile(templateUri.fsPath, { encoding: "utf8" })).toString();
```

Comment:
>readFile [](start = 30, length = 8)

Would it make sense to read only to the limit we want to look at the file, i.e. 4K instead of reading the entire file?

Reply:
DeploymentTemplate needs to full file. Also, user opened it, so it's not a concern like when reading a bunch of random files of who knows what length and that may not be intended for this use.